pkgs/build-support/fetchnodemodules/default.nix

```diff
@@ -0,0 +1,101 @@
+{ lib, stdenvNoCC, makeWrapper, nodejs }:
+
+{ src
+, hash ? ""
+, sha256 ? ""
+, name ? null
+, production ? true
+, runScripts ? false
+, makeTarball ? true
+, preferLocalBuild ? true
+, exposeBin ? false
+, npmFlags ? ""
+, ...
+} @ args:
+
+assert makeTarball -> !exposeBin;
+let
+  suffix = if makeTarball then ".tar.gz" else "";
+  name_ = if name == null then "${src.name}-node_modules${suffix}" else name;
+  nodejs_ = if args ? nodejs then args.nodejs else nodejs;
+
+  hash_ =
+    if hash != "" then { outputHashAlgo = null; outputHash = hash; }
+    else if sha256 != "" then { outputHashAlgo = "sha256"; outputHash = sha256; }
+    else throw "fetchNodeModules requires a hash for ${name_}";
+
+  customArgs = [ "name" "hash" "sha256" ];
+in stdenvNoCC.mkDerivation ({
+  inherit src production runScripts makeTarball preferLocalBuild;
+
+  name = name_;
+  nativeBuildInputs = [ nodejs_ makeWrapper ];
+
+  phases = "unpackPhase patchPhase buildPhase installPhase";
+
+  buildPhase = ''
+    if [[ ! -f package.json ]]; then
+        echo
+        echo "ERROR: The package.json file doesn't exist"
+        echo
+        exit 1
+    fi
+    if [[ ! -f package-lock.json ]]; then
```

Comment:
I think npm also supports yarn.lock as well: https://blog.npmjs.org/post/621733939456933888/npm-v7-series-why-keep-package-lockjson.html

Reply:
It does with the normal `npm i` which will automatically generate a `package-lock.json` from `yarn.lock`, but not the deterministic `npm ci` which we are using which refuses to perform any change to the lock files.